### ISSUE TITLE
Regression in 4.0.0-beta.2: script urls no longer point to publicPath

### ISSUE THREAD

**Issue by ahettlin** [jantimon/html-webpack-plugin#1098]
I have upgraded from v3.2.0 to v4.0.0-beta.2 in order to get injection of bundles generated via splitChunks to work as suggested in other issues.  However, it appears that the behavior of the plugin has changed with regard to how the urls for injected scripts are generated.  Before, it seemed that the urls were just `publicPath + bundleName.js` (I have not looked at the source code of either version yet). With the new version, the url is now a relative path from the template location to the js file, which is incorrect when the file is served by our server.

Our urls changed from 
```html
<script type="text/javascript" src="app/main/ccf-app.js"></script>
```
to
```html
<script src="../../webapp/app/main/ccf-app.js"></script>
```

I was unable to find any documentation on how to customize the url, if there is any.  I realize this is a beta version. Is this intentional? Is there any way to get the old behavior back?

### Config

```js
module.exports = {
  entry: {
    'ccf-app':	'./src/main/javascript/main/main.ts'
  },
  output: {
    path: helpers.root('src/main/webapp/app/main'),
    publicPath: 'app/main/',
    filename: '[name].[hash].js',
    chunkFilename: '[name].[chunkhash].chunk.js'
  },
  module: {
    rules : [...]
  }
  plugins: [
    new HtmlWebpackPlugin({
      template : './src/main/templates/index.ejs',
      filename: '../../../resources/templates/index.ftl',
      inject: 'body',
      chunksSortMode: 'none',
      minify: false
    }),
    ...
  ],
  optimization: {
    removeAvailableModules: false,
    removeEmptyChunks: true,
    splitChunks: {
      chunks: 'all'
    }
  }
}
```

## Environment

Node.js: 10.13.0
npm: 6.4.1
webpack: 4.25.0
html-webpack-plugin: 4.0.0-beta.2
OS: win32 10.0.17758

**Comment by Wind4** [comment]
In the v4 version, if `webpackConfig.output.publicPath` is set to a relative path, it will resolve to the relative path of entry, otherwise to `publicPath + bundleName.js`.

Can you set `webpackConfig.output.publicPath = '/app/main/'`?

**Comment by jantimon** [comment]
Otherwise you would have to copy the result from the dist folder to the correct location after generating the html

**Comment by ahettlin** [comment]
The problem with adding that leading slash is that it turns into an absolute url relative to the domain root.  Our applications run in parallel with others at our organization on the same domain.  The various applications are separated by a url path segment generated by our server (in our case, Tomcat).  That is why we need the resource url to be relative to the url of the index page, which is not the same as how the files are relative to each other on the file system. Additionally, since this path segment is virtual and doesn't represent a real folder, there is nowhere to copy the resources to.

**Comment by jantimon** [comment]
The html-webpack-plugin tries to allow relative and absolute files.  
It is using a combination of the html-webpack-plugin `filename` option and the `outputPath`.

https://github.com/jantimon/html-webpack-plugin/blob/e2276195d2f919d68d0cffc5c0a7208bf27b010c/index.js#L563-L572

Lets skip the absolute for know and take a look what the relative resolution code does.
To make it a little bit easier take a look at the following example structure:

```
├── html
│   └── index.html
└── js
    └── main.js
```

A valid relative path for the html file to work according to the output options would be:

`<script src="../js/main.js></script>`

However the way you are deploying your html files is very different from your webpack build structure and therefore this approach won't work.

Right now the public path is not taken in account anymore because of https://github.com/jantimon/html-webpack-plugin/pull/1066

But maybe that's a bug.

How do you think should the public path generation work?

**Comment by ahettlin** [comment]
I suppose the heart of the issue is that there are two kinds of relative URLs.  (relative directory traversal, which you have working, and path-relative, which is what we need).  While it may be a unique case, I can't imagine we are the only ones that need path-relative urls for our resources. And in any case, I think it would be good if the plugin supported all URL forms.  

I understand the design decision to use absolute URLs as-is and to make sure relative URLs match the file system.  However that breaks down when you are using server technologies where the URL path segments don't necessarily map up to file system folders. Apologies if I'm being redundant.

I suppose in the end, I don't have the solution. Neither myself nor my team are experts in the workings of webpack or this plugin.  We just need to keep the ability to have path-relative URLs or otherwise we will have to find a different tool. It would be great if we could keep using the html-webpack-plugin as it has been awfully convenient.  Even if the old behavior (essentially using the publicPath as-is and prepend it to the resource file name) could be hidden behind an option passed to the plugin, that would satisfy us.

**Comment by jantimon** [comment]
The easiest fix would be to configure webpack in a way that your url schema matches your file structure (this would also be the best solution if you have a webpack dev server running for development).

Once that works you can use a different plugin to copy the files to the correct location so that your backend system works as expected.

I know that this is very different from your current setup however it would also be easier to understand for the next developer if the dist folder mirrors the web server url structure.

Another solution might be to use a html-webpack-plugin hook so you can change the url of every tag very easily to match your requirements.

**Comment by ahettlin** [comment]
The way our web servers are set up with Tomcat, it is impossible to have the URL structure for static resources match the file system if you have to start from the root/domain.  Additionally, the paths change depending on whether the application is running in one of several test environments or the production environment. 

I took a quick look at the documentation for the hooks. That seems like something we could get working eventually.  It still feels like it shouldn't be this much trouble to get path-relative URLs working, especially if it's something that works in the current version of the plugin.  However, if this will be our only route when the new version is final, I suppose it's what we have to do.

**Comment by jinasonlin** [comment]
https://github.com/jantimon/html-webpack-plugin/blob/7095cd66bdb6b68c00abf521dcf11d4547cb9475/index.js#L563-L568

### Config

```js
module.exports = {
  entry: 'app.js',
  output: {
    path: 'dist',
    filename: 'js/[name].js'
    publicPath: '../../static'
  },
  plugins: [
    new HtmlWebpackPlugin(),
    ...
  ]
}
```

#### Current

`js/app.js`

#### Expected

`../../static/js/app.js`

I think we should use `typeof path === 'undefined' || path === ''` instead of `isAbsolutePath`. Relative path is needed.

**Comment by jantimon** [comment]
It sounds like merging #1066 was a mistake - could you please prepare a pull request to fix the relative path handling?

**Comment by yanni4night** [comment]
Please publish a new version to fix this

**Comment by jimmydief** [comment]
+1, was about to file my own issue for this and would appreciate a release.

**Comment by jantimon** [comment]
Released as html-webpack-plugin@4.0.0-beta.5

**Comment by ahettlin** [comment]
Thank you. I haven't had a chance to test the new version myself, but a cursory glance at the commit looks promising.

**Comment by jantimon** [comment]
I am closing this as fixed - if you have any problems please open a new issue
